### PR TITLE
doc: contribute: Rework the process for external code inclusion

### PR DESCRIPTION
**Pull request by carlescufi**
As per the agreements in the Process Group Meeting, rework the procedure
to include external code so that it reflects the reality of Zehyr today
with the choice between modules and integration into the main tree, and
decouple the non-Apache v2.0 license side of the question from the
process from importing 3rd-party code.

[ext.pdf](https://github.com/zephyrproject-rtos/zephyr/files/6190875/ext.pdf)

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

### CODE REVIEW

File: doc/guides/modules.rst [hunk 1]
```diff
@@ -842,6 +839,7 @@ Note that multiple and dependent changes to different modules can be submitted
 using exactly the same process. In this case you will change multiple entries of
 all modules that have a pull request against them.
 
+.. _submitting_new_modules:
```
Comment (carlescufi) [inline doc/guides/modules.rst:842]: Maybe we should move this section to the contribution guidelines instead

File: doc/contribute/external.rst [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+   - If rejected by the TSC, a TSC member will communicate this to
+     the contributor and the PR will be closed.
+
+   - If approved by the TSC, the TSC chair will forward the following
+     information to the Zephyr governing board for further review:
+
+     - URL for the project where the code was imported from 
+     - License under which the imported source code was released
+
```
Comment (carlescufi) [inline doc/contribute/external.rst:127]: Add the reasons for inclusion of this code

File: doc/contribute/external.rst [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+The first way to integrate external source code into the project is to simply
+import the source code files into the main ``zephyr`` repository. This
+automatically implies that the imported source code becomes part of the
+"mainline" codebase, which in turn requires that:
+
+- The code is formatted according to the Zephyr :ref:`coding style`
+- The code adheres to the project's :ref:`coding guidelines`
+- The code is subject to the same checks and verification requirements as the
+  rest of the code in the main tree, including static analysis
+
```
Comment (carlescufi) [inline doc/contribute/external.rst:83]: Add an SPDX tag if not present.
Also add an entry here: https://docs.zephyrproject.org/latest/LICENSING.html#zephyr-licensing

**Comment by kestewart** [comment]
Before accepting a module and major subsystem,  the maintainer should be identified.   Can this be added to any policy checklist.

**Review by kestewart — COMMENTED** [review]
Please determine what should happen in the MAINTAINER file when a significant module or subsystem is added into the Zephyr repo.

File: .github/ISSUE_TEMPLATE/ext-source.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+Will the Zephyr project have a direct dependency on the component, or
+will it be included via an abstraction layer with this component as a
+replaceable implementation?
+
+## Revision
+
+Version or SHA you would like to integrate initially
+
+## License
```
Comment (galak) [inline .github/ISSUE_TEMPLATE/ext-source.md:57]: Do you want to make any comment here about non-OSI licenses / permissive licenses would be rejected?

File: .github/ISSUE_TEMPLATE/ext-source.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+## Mode of integration
+
+Describe whether you'd like to integrate this exernal component in the main tree
+or as a module, and why. If the mode of integration is a module, suggest a
+repository name for the module.
+
+## Pull Request
+
+Pull request (if any) with the actual implementation of the integration, be it
```
Comment (ioannisg) [inline .github/ISSUE_TEMPLATE/ext-source.md:27]: DNM

File: doc/contribute/external.rst [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+- All files contain an SPDX tag if not already present
+- An entry is added to the `licensing page <zephyr_licensing>`
+
+This mode of integration can be applicable to both small and large external
+codebases, but it is typically used more commonly with the former.
+
+Integration as a module
+-----------------------
+
+The second way of integrating external source code into the project is to import
```
Comment (ioannisg) [inline doc/contribute/external.rst:92]: Maintaining a module

File: doc/contribute/external.rst [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+into the project:
+
+- Is this the most optimal way to introduce the functionality to the project?
+  Both the cost of implementing this internally and the one incurred in
+  maintaining an externally developed codebase need to be evaluated.
+- Is the external project being actively maintained? This is particularly
+  important for source code that deals with security or cryptography.
+- Have alternatives to the particular implementation proposed been considered?
+  Are there other open source project that implement the same functionality?
+
```
Comment (galak) [inline doc/contribute/external.rst:63]: Anything about maintenance?

File: doc/contribute/external.rst [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+All contributions to the Zephyr project are submitted through GitHub
+issues and pull requests (PR) following the Zephyr Project's
+:ref:`Contribution workflow`.
+
+Before external source code is accepted, it must be accepted by the Technical
+Steering Committee (TSC) as suitable and appropriate. Depending on whether the
+imported source code needs to be integrated in the main tree or else as a
+module, the procedure differs slightly, as outlined below.
+
+A request for external source code integration should be initialized using an
```
Comment (ioannisg) [inline doc/contribute/external.rst:113]: not needed, remove

File: doc/contribute/external.rst [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+RFC issue in the Zephyr project issue tracking system on GitHub with details
+about the source code and how it integrates into the project.
+
+Follow the steps below to request a new module:
```
Comment (ioannisg) [inline doc/contribute/external.rst:117]: request integration of external code

File: .github/ISSUE_TEMPLATE/ext-source.md (hunk 4 above)
Comment (carlescufi) [reply .github/ISSUE_TEMPLATE/ext-source.md:57]: No, let's leave that for the main doc instead.

**Review by nashif — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Comment by carlescufi** [comment]
@MaureenHelm @galak @kestewart @ioannisg this should be ready for re-review now. I've attached a rendered PDF in the PR description.

File: doc/contribute/external.rst [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+process.
+
+.. _external-src-process:
+
+Submission and review process
+=============================
+
+Before external source code can be included in the project, it must be reviewed
+and accepted by the Technical Steering Committee (TSC) and, in some cases, by
+the Zephyr Board of Directors.
```
Comment (MaureenHelm) [inline doc/contribute/external.rst:116]:
```suggestion
the Zephyr Governing Board.
```

**Comment by carlescufi** [comment]
@MaureenHelm thank you, fixed.

File: doc/contribute/external.rst [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+   External source code licensed under the Apache-2.0 license is not subject to
+   this section.
+
+Integrating code into the Zephyr Project from other projects that use a license
+other than the Apache 2.0 license needs to be fully understood in
+context and approved by the `Zephyr governing board`_, as described in the
+`Zephyr project charter`_. The board will automatically reject licenses that
+have not been approved by the `Open Source Initiative (OSI)`_. See the
+:ref:`external-src-process` section for more details.
```
Comment (ioannisg) [inline doc/contribute/external.rst:30]: Note: do we need to care if this concerns content that does not end up in the firmware built?

**Review by galak — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by gregshue — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Comment by ioannisg** [comment]
@MaureenHelm we briefly reviewed this in yesterday's Process meeting, and the consensus is that it is merge-able, now. Please review when you have time.

**Comment by carlescufi** [comment]
@ioannisg no, the one that needs review is https://github.com/zephyrproject-rtos/zephyr/pull/33299. This one is just a basic cleanup of the doc files, merging.

**Pull request merged**